### PR TITLE
support member injection for superclasses

### LINKED ISSUE #343: Member injection does not account for superclass injections
Given this example:

```kotlin
class DebugExperimentsAppActivity : DebugExperimentsBaseActivity() {

  @Inject
  @ForScope(AppScope::class)
  lateinit var featureFlagStore: FeatureFlagStore
}

abstract class DebugExperimentsBaseActivity : DebugBaseActivity() {

  @Inject
  lateinit var experimentManager: ExperimentManager

  @Inject
  lateinit var darkModeHelper: DarkModeHelper
}
```

anvil will only generate injection for members in the injected class

```kotlin
public class DebugExperimentsAppActivity_MembersInjector(
  private val param0: Provider<FeatureFlagStore>
) : MembersInjector<DebugExperimentsAppActivity> {
  public override fun injectMembers(instance: DebugExperimentsAppActivity): Unit {
    injectFeatureFlagStore(instance, param0.get())
  }

  public companion object {
    @JvmStatic
    public fun create(param0: Provider<FeatureFlagStore>):
        DebugExperimentsAppActivity_MembersInjector =
        DebugExperimentsAppActivity_MembersInjector(param0)

    @JvmStatic
    @InjectedFieldSignature("...")
    @ForScope(value = AppScope::class)
    public fun injectFeatureFlagStore(instance: DebugExperimentsAppActivity,
        featureFlagStore: FeatureFlagStore): Unit {
      instance.featureFlagStore = featureFlagStore
    }
  }
}
```

But dagger expects this injector to inject _all_ inherited members too

```java
this.debugExperimentsAppActivityMembersInjectorProvider = InstanceFactory.create(DebugExperimentsAppActivity_MembersInjector.create(provideExperimentManagerProvider, provideDarkModeHelperProvider, provideFeatureFlagStoreProvider));
```

### PR DESCRIPTION
**Pull request by RBusarow**
fixes #343

Basically, these changes use PSI to find superclasses by parsing the FqName from a `KtClassOrObject`'s supertype list, then using that to look up the super's `KtClassOrObject` from `RealAnvilModuleDescriptor`.  When this method no longer returns results, that means any additional super-types come from other modules/dependencies and it's safe to use descriptors.

This required forking a lot of the psi-based parsing for `Parameter` for use with Descriptor types.  

I also turned `Parameter` into an interface and created `MemberInjectParameter`, which is the same thing except for a few extra fields for member-injected specific stuff.  I don't love the names I chose, and there are a few options I'm on the fence about, like using interface delegation for MemberInjectedParameter and losing the data modifier.

**`Parameter` naming is changed to use original names**.  In order to support collecting all the parameters from different sources, I had to either pass around a `startIndex` parameter, or use original parameter names.  Continuing to use the index seems error-prone, and the Kotlin compiler should already be ensuring that the original parameter names are unique.  There may be an edge case I'm not considering, but all tests pass.

I believe this approach of doing a reverse lookup for the `KtClassOrObject` could also be used to solve the `AssistedFactory` generation issue (#326) without a major rewrite, but I haven't fully tested this yet.

TODO
- [x] Add superclass injection into MemberInjectors
- [x] Parse Qualifier annotations.

### CODE REVIEW

**Review by vRallev — APPROVED** [review]
Unfortunately, I don't believe there's a good way to avoid the duplication that you called out. PSI and descritptor APIs are very different and I don't believe it makes sense to hide this behind common interfaces / types. The PR is awesome, as usual. That was a big lift.

File: compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt [hunk 1]
```diff
@@ -185,6 +186,36 @@ public inline fun <reified T> KtAnnotationEntry.findAnnotationArgument(
     }
 }
 
+/**
+ * This will return all super classes which are parsed as PSI, which means anything defined in this
+ * module. This will include generated code, assuming it has already been generated. It **will not**
+ * include any classes from other modules or external dependencies.
+ *
+ * The first element in the returned list represents the direct superclass to the receiver. The last
+ * element represents the class which is furthest up-stream.
```
[the remark this answers is not in the document]
Comment (RBusarow) [reply compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/PsiUtils.kt:195]: It *could*, but I'm filtering them out for this function.  This function could be duplicated (`allPsiSuperTypes()`?) without the in order to solve the other resolution problems.

File: compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt [hunk 2]
```diff
@@ -183,7 +183,7 @@ public fun Any.invokeGet(vararg args: Any?): Any {
 
 @ExperimentalAnvilApi
 public fun Any.getPropertyValue(name: String): Any {
-  return this::class.java.getDeclaredField(name).use { it.get(this) }
+  return this::class.java.fields.first { it.name == name }.use { it.get(this) }
```
[the remark this answers is not in the document]
Comment (RBusarow) [reply compiler-utils/src/testFixtures/java/com/squareup/anvil/compiler/internal/testing/TestUtils.kt:186]: "Declared" fields don't include fields declared in super-classes.  So it didn't break, but it didn't extend to the new requirements.

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/RealAnvilModuleDescriptor.kt [hunk 3]
```diff
@@ -49,6 +49,11 @@ class RealAnvilModuleDescriptor(
       ?.fqName
   }
 
+  override fun getKtClassOrObjectOrNull(fqName: FqName): KtClassOrObject? {
+    return allClasses
+      .firstOrNull { it.fqName == fqName }
+  }
```
[the remark this answers is not in the document]
Comment (RBusarow) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/RealAnvilModuleDescriptor.kt:55]: Nice!

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/TestClasses.kt [hunk 4]
```diff
@@ -13,3 +16,23 @@ abstract class OuterClass constructor(
...
+annotation class ClassQualifier(val value: KClass<*>)
+
+@Qualifier
+annotation class EnumArrayQualifier(val value: Array<Size>)
+
+@Qualifier
+annotation class EnumQualifier(val value: Size)
+
+@Qualifier
+annotation class IntQualifier(val value: Int)
```
[the remark this answers is not in the document]
Comment (RBusarow) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/TestClasses.kt]: Yeah, I felt pretty dirty doing it...  Fixed!

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/Parameter.kt [hunk 5]
```diff
@@ -0,0 +1,74 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.squareup.anvil.compiler.codegen.Parameter.AssistedParameterKey
+import com.squareup.kotlinpoet.AnnotationSpec
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.ParameterizedTypeName
+import com.squareup.kotlinpoet.TypeName
+import org.jetbrains.kotlin.name.FqName
+
+sealed interface Parameter {
```
[the remark this answers is not in the document]
Comment (RBusarow) [reply compiler/src/main/java/com/squareup/anvil/compiler/codegen/Parameter.kt:10]: Thanks for the thanks!

I wasn't sure about making this sealed.  I'm still not.
